src/whirligig.js

```diff
@@ -1,67 +1,73 @@
-  values,
-  includes,
-  isWhollyInView
+  onSwipe,
+  values
 } from './utils'
-import { bool, number, string, func, array, oneOfType, object, node } from 'prop-types'
-// const tap = (msg) => (thing) => { console.log(msg, thing); return thing }
-const wrapAroundValue = (val, max) => ((val % max) + max) % max
-const hardBoundedValue = (val, max) => Math.max(0, Math.min(max, val))
-const normalizeIndex = (idx, len, wrap = false) => wrap ? wrapAroundValue(idx, len) : hardBoundedValue(idx, len - 1)
-
-export default class Whirligig extends Component {
-  static propTypes = {
-    afterSlide: func,
-    animationDuration: number,
-    beforeSlide: func,
-    children: oneOfType([node, array, string]),
-    className: string,
-    easing: func,
-    infinite: bool,
-    innerRef: func,
-    nextKeys: array,
-    prevKeys: array,
-    preventAutoCorrect: bool,
-    preventScroll: bool,
-    preventSwipe: bool,
-    onSlideClick: func,
-    snapToSlide: bool,
-    slideTo: number,
-    slideBy: number,
-    slideClass: oneOfType([array, string, object]),
-    startAt: number,
-    style: object,
-    gutter: (props, propName, componentName) => {
-      const prop = props[propName]
-      if (
-        typeof parseInt(prop, 10) === 'number' &&
-        !isNaN(Number(prop))
-      ) {
-        return new Error(`Invalid value (${prop}) of prop '${propName}' supplied to ${componentName}.
-        The value of ${propName} should be a valid css length unit
-        (https://developer.mozilla.org/en-US/docs/Web/CSS/length).`)
-      }
-    },
-    visibleSlides: number
-  }
+
+export type WhirligigProps = {
+  afterSlide: (any) => any,
+  animationDuration: number,
+  beforeSlide: (any) => any,
+  children: any[] | string | React.Node,
+  className: string,
+  easing: (number) => number,
+  gutter: string,
+  infinite: bool,
+  innerRef: (any) => void,
+  nextKeys: string[],
+  onSlideClick: (any) => void,
+  prevKeys: string[],
+  preventAutoCorrect: bool,
+  preventScroll: bool,
+  preventSwipe: bool,
+  slideBy: number,
+  slideClass: string,
+  slideTo: number,
+  snapToSlide: bool,
+  startAt: number,
+  style: Object,
+  visibleSlides: number
+}
+
+export type WhirligigState = {
+  activeIndex: number,
+  isAnimating: bool,
+  slideBy: number,
+  visibleSlides: number
+}
+
+export default class Whirligig extends React.Component<WhirligigProps, WhirligigState> {
+  DOMNode: Element
+  canSelfCorrect: bool
+  childCount: number
+  eventListeners: any[]
+  isInteracting: () => bool
+  isScrolling: bool
+  whirligig: any
+  next: (any) => Promise<*>
+  prev: (any) => Promise<*>
+  slideTo: (number, ?{ immediate: bool }) => Promise<*>
 
   static defaultProps = {
-    afterSlide: () => {},
+    afterSlide: noop,
     animationDuration: 500,
-    beforeSlide: () => {},
+    beforeSlide: noop,
     gutter: '1em',
-    nextKeys: ['ArrowRight'],
-    prevKeys: ['ArrowLeft'],
+    nextKeys: [ 'ArrowRight' ],
+    prevKeys: [ 'ArrowLeft' ],
```

Comment:
are you sure there should be spaces there?

Reply:
It's preference.

Reply:
okay. I thought we had some lint rules against it